cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,549 @@
+                "--system-mounts", "--system-resolvconf",
+                new_img_path,
+                "--",
+                "sh",
+                "-c",
+                f"chroot ${{MOUNTPOINT}} apt-get update; DEBIAN_FRONTEND=noninteractive chroot ${{MOUNTPOINT}} apt-get install -o  Dpkg::Options::='--force-confold' dhcpcd5 -y --force-yes",
+            ],
+            [
+                "sudo",
+                "mount-image-callback",
+                new_img_path,
+                "--",
+                "sh",
+                "-c",
+                f"cp {deb_path} ${{MOUNTPOINT}}/.; chroot ${{MOUNTPOINT}} dpkg -i /{os.path.basename(deb_path)}",
+            ],
+        ]
+    elif deb_path.startswith("ppa:"):
+        commands = [
+            [
+                "sudo",
+                "mount-image-callback",
+                "--system-mounts", "--system-resolvconf",
+                new_img_path,
+                "--",
+                "sh",
+                "-c",
+                f"chroot ${{MOUNTPOINT}} add-apt-repository {deb_path} -y; DEBIAN_FRONTEND=noninteractive chroot ${{MOUNTPOINT}} apt-get install -o  Dpkg::Options::='--force-confold' cloud-init -y",
+            ],
+        ]
+    else:
+        raise RuntimeError(
+            f"Invalid deb_path provided: {deb_path}. Expected local .deb or" " ppa:"
+        )
+    for command in commands:
+        logging.debug(f"--- Running: {' '.join(command)}")
+        subprocess.check_call(command)
+    return new_img_path
+
+
+def collect_bootspeed_data(instance):
+    """Collect and process bootspeed data from the instance."""
+    start_time = time.time()
+    data = {"time_at_ssh": retry_cmd(instance, "date --utc +'%b %d %H:%M:%S.%N'")}
+    ssh_time = time.time()
+    data["cloudinit_status"] = json.loads(
+        instance.execute("cloud-init status --format=json --wait")
+    )
+    cloudinit_done_time = time.time()
+    data["client_time_to_ssh"] = ssh_time - start_time
+    data["client_time_to_cloudinit_done"] = cloudinit_done_time - start_time
+    data["image_builddate"] = retry_cmd(
+        instance, "grep serial /etc/cloud/build.info"
+    ).split()[-1]
+    data["systemd_analyze"] = retry_cmd(instance, "systemd-analyze")
+    data["systemd_analyze_blame"] = retry_cmd(instance, "systemd-analyze blame")
+    data["cloudinit_version"]: retry_cmd(instance, "dpkg-query -W cloud-init").split()[
+        1
+    ]
+    data["cloudinit_analyze"] = instance.execute("cloud-init analyze show")
+    data["cloudinit_analyze_blame"] = instance.execute("cloud-init analyze blame")
+
+    for service in (
+        "cloud-init.service",
+        "cloud-init-local.sevice",
+        "cloud-config.service",
+        "cloud-final.service",
+        "systemd-user-sessions.service",  # Affects sytem login time
+        "snapd.seeded.service",  # Has impact on cloud-config.service
+    ):
+        data[
+            f"criticalchain_{service.replace('-', '_').replace('.', '_')}"
```

Comment:
In some cases this tool turns unstructured data (output of `cloud-init blame`) into structured output, which may be helpful for any data analysis tool. In this case, however, the data is gathered, but not processed into structured output. It would be useful to gather this data in structured output as well.